### PR TITLE
perl-packages.nix updated to actual CPAN versions [staging]

### PR DESCRIPTION
**Pull request by volth**
###### Motivation for this change

Automatically update perlPackages 
Fixes https://github.com/NixOS/nixpkgs/issues/35693

### CODE REVIEW

**Comment by Mic92** [comment]
@vcunat Should this be its own jobset in hydra?
@volth What you have tested so far regarding your updates?

**Comment by Mic92** [comment]
2018 will be the year, where machines take nixpkgs over.

**Comment by volth** [comment]
> What you have tested so far regarding your updates?

very little, the majority of those packets are not used in my closures. and only on ```x86_64-linux```
so yes, it should be run on Hydra with all the tests

**Comment by Mic92** [comment]
@volth did you only update those, that do not have new dependencies?

**Comment by volth** [comment]
> did you only update those, that do not have new dependencies?

no, matching ```nixpkgs``` dependency graph against CPAN's ([example](wget https://cpan.metacpan.org/authors/id/E/ET/ETHER/B-Hooks-OP-Check-0.22.tar.gz.meta)) would be a nightmare.
Ww will see if some of the builds fail due to lack of new dependencies.

**Comment by volth** [comment]
@Mic92 can you run it on Hydra or the CI bot?
It is not as big job as it seems, most of the perl libs are small and written in plain C

**Comment by Mic92** [comment]
@volth I can merge it into staging, but I have not yet the `Super Cow Powers` to build staging immediately.
I feel for this pull request it might be interesting to not having it interfere with other big rebuilds in staging,
hence a dedicated jobset would be helpful.

**Comment by volth** [comment]
oops, there is already script by eelco: https://github.com/NixOS/nixpkgs/blob/master/maintainers/scripts/nix-generate-from-cpan.pl
just nobody run in many years

**Comment by volth** [comment]
I have fixed deps in almost every place where they was broken (with the exception of few packages which requires too long build, like wxwidgets, they are still untested). You are right, that should be automated too.

Some unit tests appeared in the new versions of the packages do not pass. They are to be either fixed or disabled.

**Comment by Mic92** [comment]
The stuff we usually commit to staging is supposed to build/work on its own, but may break the build of other packages. Is it possible to extract a subset of packages that builds (and hopefully works)?

**Comment by volth** [comment]
That makes sense.

Everything must work as by now.

Not checked only ```GSSAPI``` because ```heimdal``` (the C library in depends on) is broken on staging

**Comment by volth** [comment]
oh, already has merge conflict

**Comment by Mic92** [comment]
@FRidh can you create a jobset for this package set update?

File: pkgs/top-level/perl-packages.nix [hunk 1]
```diff
@@ -389,6 +398,15 @@ let self = _self // overrides; _self = with self; {
     buildInputs = [ TestNoWarnings Moo TypeTiny ];
   };
 
+  ArrayDiff = buildPerlPackage rec {
+    name = "Array-Diff-0.07.tar.gz";
```
Comment (rycee) [inline pkgs/top-level/perl-packages.nix]: This name is bad. The `nix-generate-from-cpan Array::Diff` produces

```nix
  ArrayDiff = buildPerlPackage rec {
    name = "Array-Diff-0.07";
    src = fetchurl {
      url = "mirror://cpan/authors/id/T/TY/TYPESTER/${name}.tar.gz";
      sha256 = "78e7d3a3374c74e117b7198f8e6649367a2906561b8a485d3870f242fa8f8346";
    };
    propagatedBuildInputs = [ AlgorithmDiff ClassAccessor ];
    meta = {
      description = "Find the differences between two arrays";
      license = with stdenv.lib.licenses; [ artistic1 gpl1Plus ];
    };
  };
```

File: pkgs/top-level/perl-packages.nix [hunk 2]
```diff
@@ -857,6 +875,15 @@ let self = _self // overrides; _self = with self; {
     };
   };
 
+  BUtils = buildPerlPackage {
```
Comment (rycee) [inline pkgs/top-level/perl-packages.nix]: From `nix-generate-from-cpan B::Utils`:

```nix
  BUtils = buildPerlPackage rec {
    name = "B-Utils-0.27";
    src = fetchurl {
      url = "mirror://cpan/authors/id/E/ET/ETHER/${name}.tar.gz";
      sha256 = "f97f53f6a3050109aa414feeb184cad101812d41760e952b5d84993f6685ffea";
    };
    buildInputs = [ ExtUtilsDepends ];
    propagatedBuildInputs = [ TaskWeaken ];
    meta = {
      homepage = http://search.cpan.org/dist/B-Utils;
      description = "Helper functions for op tree manipulation";
      license = with stdenv.lib.licenses; [ artistic1 gpl1Plus ];
    };
  };
```

File: pkgs/top-level/perl-packages.nix [hunk 3]
```diff
@@ -2152,12 +2180,21 @@ let self = _self // overrides; _self = with self; {
     };
   };
 
+  ClassTiny = buildPerlPackage {
```
Comment (rycee) [inline pkgs/top-level/perl-packages.nix]: From `nix-generate-from-cpan Class::Tiny`:

```nix
  ClassTiny = buildPerlPackage rec {
    name = "Class-Tiny-1.006";
    src = fetchurl {
      url = "mirror://cpan/authors/id/D/DA/DAGOLDEN/${name}.tar.gz";
      sha256 = "2efcbd31528be51d3022c616768558b78c6172df5f03c5dc698939f65488cb4e";
    };
    meta = {
      homepage = https://github.com/dagolden/Class-Tiny;
      description = "Minimalist class construction";
      license = stdenv.lib.licenses.asl20;
    };
  };
```

File: pkgs/top-level/perl-packages.nix [hunk 4]
```diff
@@ -2218,6 +2255,28 @@ let self = _self // overrides; _self = with self; {
...
+      url = "mirror://cpan/authors/id/H/HE/HERMES/Clone-Choose-0.008.tar.gz";
+      sha256 = "0grf0bg1ldzqakaq7aaigjngwd07zybrlmwjcnb0q986q0jcfank";
+    };
+    meta = {
+      description = "Choose appropriate clone utility";
+      license = with stdenv.lib.licenses; [ artistic1 gpl1Plus ];
+    };
+  };
+
+  CodeTidyAll = buildPerlPackage rec {
```
Comment (rycee) [inline pkgs/top-level/perl-packages.nix]: From `nix-generate-from-cpan Code::TidyAll`:

```nix
  CodeTidyAll = buildPerlPackage rec {
    name = "Code-TidyAll-0.69";
    src = fetchurl {
      url = "mirror://cpan/authors/id/D/DR/DROLSKY/${name}.tar.gz";
      sha256 = "13fa4f7ac6ed7ea514072957df8eaae5ee349f5f1def9c6a988a5197b0bd855e";
    };
    buildInputs = [ TestClassMost TestDifferences TestFatal TestWarnings librelative ];
    propagatedBuildInputs = [ CaptureTiny ConfigINI FileWhich FileZglob Filepushd IPCRun3 IPCSystemSimple ListCompare ListSomeUtils LogAny ModuleRuntime Moo PathTiny ScopeGuard Specio SpecioLibraryPathTiny TextDiff TimeDate TimeDurationParse TryTiny ];
    meta = {
      homepage = http://metacpan.org/release/Code-TidyAll;
      description = "Engine for tidyall, your all-in-one code tidier and validator";
      license = with stdenv.lib.licenses; [ artistic1 gpl1Plus ];
    };
  };
```

File: pkgs/top-level/perl-packages.nix [hunk 5]
```diff
@@ -2557,6 +2616,14 @@ let self = _self // overrides; _self = with self; {
     };
   };
 
+  CPANDistnameInfo = buildPerlPackage rec {
```
Comment (rycee) [inline pkgs/top-level/perl-packages.nix]: From `nix-generate-from-cpan CPAN::DistnameInfo`:

```nix
  CPANDistnameInfo = buildPerlPackage rec {
    name = "CPAN-DistnameInfo-0.12";
    src = fetchurl {
      url = "mirror://cpan/authors/id/G/GB/GBARR/${name}.tar.gz";
      sha256 = "2f24fbe9f7eeacbc269d35fc61618322fc17be499ee0cd9018f370934a9f2435";
    };
    meta = {
      description = "Extract distribution name and version from a distribution filename";
      license = with stdenv.lib.licenses; [ artistic1 gpl1Plus ];
    };
  };
```

File: pkgs/top-level/perl-packages.nix [hunk 6]
```diff
@@ -2929,6 +2996,14 @@ let self = _self // overrides; _self = with self; {
     };
   };
 
+  curry = buildPerlPackage {
```
Comment (rycee) [inline pkgs/top-level/perl-packages.nix]: From `nix-generate-from-cpan curry`:

```nix
  curry = buildPerlPackage rec {
    name = "curry-1.001000";
    src = fetchurl {
      url = "mirror://cpan/authors/id/M/MS/MSTROUT/${name}.tar.gz";
      sha256 = "8b186079202064fb3dcc25cf7290aa3f5d551d98d1b86d5844706a760c1f56d4";
    };
    meta = {
      description = "Create automatic curried method call closures for any class or object";
      license = with stdenv.lib.licenses; [ artistic1 gpl1Plus ];
    };
  };
```

**Comment by rycee** [comment]
Overall I like this a lot, many thanks for taking the time to clean up this package set!

I notice some of the new packages seem to contain mistakes, in particular having `tar.gz` in the name, e.g., `name = "CPAN-DistnameInfo-0.12.tar.gz";` and missing `meta` fields, in particular missing licensing information. `nix-generate-from-cpan` seem to be doing a bit better for some of these and I've included some examples (I didn't do all of them since nix-generate-from-cpan is not the fastest thing in the world :smiley:).

**Comment by volth** [comment]
yes, there was a manual cleanup to make it build.
maybe better just to close it and re-run ```nix-generate-from-cpan```

**Comment by rycee** [comment]
Unfortunately nix-generate-from-cpan doesn't do bulk updates at all. It's a very simple tool that lets you generate an expression for individual package. So I'd urge you to reopen this PR.

**Comment by volth** [comment]
Ok, I need to improve the tool

**Comment by volth** [comment]
It need to match the dependency graphs (CPAN and nixpkgs) and analyse the tarballs to chose either ```buildPerlPackage``` or ```buildPerlModule``` (based on presence/absence of files like Build.PL and Makefile)

**Comment by rycee** [comment]
The script does already do the `buildPerlPackage` and `buildPerlModule` thing:

https://github.com/NixOS/nixpkgs/blob/56904d7c423f2b13b37fbd29f39bbb4b52bc7824/maintainers/scripts/nix-generate-from-cpan.pl#L433-L434

and in my experience this works quite reliably.

I think the main difficulty is the the dependency graph thing, for example the `nix-generate-from-cpan.pl` tool will blindly emit the dependencies listed in the package description on CPAN that sometimes misses required dependencies. So historically, using the `nix-generate-from-cpan.pl` tool has involved manual checking of the generated expression and fixing potential issues with the build inputs, license field, and so on before putting it in `perl-packages.nix`…

But yeah, I think the code in `nix-generate-from-cpan.pl` could be used as a base (it'll need a bit of caching here and there to be a bit faster).

Ideally I think we'd set up the Perl package set to be similar to the Haskell one with a file that contains the automatically generated packages and then run this through an override mechanism where we can include the necessary fixes. I've several times thought about setting something like that up (something like perlPackagesNg) but since I no longer regularly work with Perl I've lost much of my motivation.

**Comment by rycee** [comment]
Wow!

**Comment by volth** [comment]
Ok, another approach by a smarter tool which reads the dependency graph from span and matches it against nixpkgs and the packages included in the oldest perl supported by nixpkgs (which is ```perl22```)

The subtle moments:
1. Some libs are not in ```perl-packages.nix``` but in external files and they are not updated only for that reason. Some other because of conditional expressions in ```buildInputs```. Either they should be updated manually or the tool should be more smart. 
2. ```Test*``` dependencies never go to ```proparatedBuildInputs``` because the propagation leads to conflicts of incompatible test frameworks (for example Test::Simple before and after 1.3)
3. There could be bugs related to version comparison, as some of libs use semantic version whilt other do not (```1.4201``` could be earlier than ```1.45```), so review is needed as the bot could downgrade a library.
4.

**Comment by Mic92** [comment]
@volth what if a Test* packages requieres a different Test* package, should this not be propagated?

**Comment by volth** [comment]
> @volth what if a Test* packages requieres a different Test* package, should this not be propagated?

currently it is handled in errata list (along with other dependency graph patches, like ones to break circular dependencies), but it definitively could be optimized by allowing some of the Test packages to propagate. I preferred to be safe for the beginning.

**Comment by volth** [comment]
repushed to avoid ```meta.description = "unknown"```

**Comment by volth** [comment]
we could try to switch to perl 5.26 after this.
```B-C``` is already broken anyway for some time (on the latest 5.24.x minor bump?) and no one cries, the rest of libraries had no big problems with perl 5.26 as far as I remember.

**Comment by Mic92** [comment]
Turns out perl-B-C is a dependency of hydra?

```console
$ nix-shell -p hydra
builder for '/nix/store/cd7inxbk85lrcfbjjvklyqjm2s8aql68-perl-B-C-1.54.drv' failed with exit code 2
cannot build derivation '/nix/store/ybfbsh5h7ig1pa0m66vgzyw5dhhymqim-perl-Sub-Name-0.21.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/r1mq2167if8l6p710g44w075rx46r7b4-perl-DBIx-Class-0.082841.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/ggqqmk4qafax51za7bblakq626n5gq8w-perl-Package-DeprecationManager-0.17.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/9i8pf1jkfsmxmgg3sfv3b6qblzwg0zdf-perl-Sub-Exporter-ForMethods-0.100052.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/xl7r3jrxw75szm9y8qy3s0lap663jdd7-perl-DBIx-Class-Schema-Loader-0.07048.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/39rw9dqq3w50ykqpxvvnajp3n8y4hcgl-perl-DateTime-Format-Strptime-1.75.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/qyjidc1lks2sgb0r7yi4i4dqwm8k4hf5-perl-Moose-2.2010.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/d7vcdlnmlcb87plxg68hqc5m9whlxidw-perl-MooseX-Types-0.50.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/5b2m6nq144azsp2ldp35yy2swlp7840r-perl-Catalyst-Model-DBIC-Schema-0.65.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/zpy6mviiv5kcaaa676pa82y9pags4mxx-perl-Catalyst-TraitFor-Request-ProxyBase-0.000005.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/37n08six0q36f8ryqqdzpjqn4fnnxc1y-perl-Data-Stream-Bulk-0.11.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/fgj4vhv9bp2nmbhjl2ddjwwpp6flxa2q-perl-Data-Visitor-0.30.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/6hy5zw6zgclnf70j2nfwrl967hvw8a0f-perl-DateTime-Format-Builder-0.81.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/pknwczslpxa8dkym6lh5lf96malglc64-perl-MooseX-Emulate-Class-Accessor-Fast-0.00903.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/za4rirpnzh1j2ahr1halfppriv7cxvbp-perl-MooseX-MethodAttributes-0.31.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/hdhpy0wqp28bl629mynv379piq28w6ss-perl-MooseX-Role-Parameterized-1.10.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/mnkvkdrdskd0fhgqfbywpa2lyxxzvcg6-perl-MooseX-Role-WithOverloading-0.17.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/kyv8br13iis9wkbs6g8pbigz0zbxn4ys-perl-MooseX-StrictConstructor-0.21.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/dp5zvbyzvpbyjl96h9ih9rzmn2k3911h-perl-MooseX-Types-DateTime-0.13.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/ij53qaf54lnvldz1vafcwac1zxbs9ypb-hydra-perl-deps.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/g3v62728vssdvh3x3h1rfgq511m3glya-hydra-2017-11-21.drv': 1 dependencies couldn't be built
error: build of '/nix/store/g3v62728vssdvh3x3h1rfgq511m3glya-hydra-2017-11-21.drv' failed
```

**Comment by volth** [comment]
It is broken on both ```master``` and ```staging```.

```
NIXPKGS_CONFIG= NIX_PATH=nixpkgs=https://github.com/NixOS/nixpkgs/archive/master.tar.gz nix-build -E 'with (import <nixpkgs> {}); perlPackages.BC'
NIXPKGS_CONFIG= NIX_PATH=nixpkgs=https://github.com/NixOS/nixpkgs/archive/staging.tar.gz nix-build -E 'with (import <nixpkgs> {}); perlPackages.BC'
```

Then ```hydra``` must be too

**Comment by volth** [comment]
Anyway, ```B-C``` and ```perl526``` are irrelevant to the PR content.
Here I tried not to touch the packages which do not pass its tests on ```staging``` (which also bloated the errata list) until their problems will be resolved

**Comment by volth** [comment]
ping me when you will be ready to run a hydra jobset or to merge to staging, I will rerun it then with actual inputs.

**Comment by Mic92** [comment]
@volth I finally get a hydra access. 
https://hydra.nixos.org/jobset/nixpkgs/cpan-update#tabs-evaluations will trigger build based on cpan-update branch on nixpkgs.

**Comment by Mic92** [comment]
I merged your branch here. Onto it so far. Please open a new pull request, if you have new changes, before the jobset is started for the first time.

**Comment by vcunat** [comment]
I switched the jobset into "enabled" state; otherwise it wouldn't evaluate automatically, I think.

**Comment by volth** [comment]
That may have sense to rebase (or better rerun ```cpan2nix```) to match the point of ```staging``` evaluation on Hydra to see if it breaks or cure something.
The ```staging``` branch this PR was based on had some perl packages broken (although they were not broken on ```master``` at that time).
If ```cpan-updates``` were based on arbitrary point of ```staging``` - not evaluated on Hydra - it would be hard to say what is broken only in ```cpan-updates``` and what is broken on ```cpan-updates``` **and** ```staging```

**Comment by shlevy** [comment]
@volth I'm confused, do you want to reopen this?

**Comment by volth** [comment]
> @volth I'm confused, do you want to reopen this?

@shlevy No, I can run the same script against actual ```master``` or ```staging```.

The relevancy of such PRs is quickly lost over time: ```perl-packages.nix``` has changes since this PR, some of the packages have been updated manually, ```Socket6``` has been fixed, etc

**Comment by shlevy** [comment]
OK, let me know when that's ready!

**Comment by volth** [comment]
@shlevy: this is ready-to-merge against current ```staging``` https://github.com/NixOS/nixpkgs/compare/staging...volth:cpan-update-20180406 (including fixes suggested by @rycee about not inserting needless ```version=...``` and strip "v" prefix from package versions)

I did not build all the ```nixpkgs```, but all the changed derivations have been built succesfuly

**Comment by shlevy** [comment]
Should I reopen this PR then?

**Comment by volth** [comment]
Only it you feel https://github.com/NixOS/nixpkgs/pull/36147/commits would be more convenient than https://github.com/NixOS/nixpkgs/compare/staging...volth:cpan-update-20180406

**Comment by shlevy** [comment]
1d87af68c1d9c7bbb6f026ab2873dc524d86efcf

**Pull request closed without merging**